indexing-service/src/test/java/org/apache/druid/indexing/overlord/TaskLockboxTest.java

```diff
@@ -1272,6 +1276,627 @@ public void testGetLockedIntervalsForEqualPriorityTask() throws Exception
     );
   }
 
+  @Test
+  public void testExclusiveLockCompatibility() throws Exception
+  {
+    final List<Task> tasks = new ArrayList<>();
+
+    final TaskLock theLock = insertTaskLock(
+        TaskLockType.EXCLUSIVE,
+        Intervals.of("2017/2018"),
+        MEDIUM_PRIORITY,
+        tasks
+    );
+    Assert.assertNotNull(theLock);
+    Assert.assertFalse(theLock.isRevoked());
+
+    // Another exclusive lock cannot be created for an overlapping interval
+    Assert.assertNull(
+        insertTaskLock(
+            TaskLockType.EXCLUSIVE,
+            Intervals.of("2017-05-01/2017-06-01"),
+            MEDIUM_PRIORITY,
+            tasks
+        )
+    );
+
+    // A shared lock cannot be created for an overlapping interval
+    Assert.assertNull(
+        insertTaskLock(
+            TaskLockType.SHARED,
+            Intervals.of("2016/2019"),
+            MEDIUM_PRIORITY,
+            tasks
+        )
+    );
+
+    // A replace lock cannot be created for an overlapping interval
+    Assert.assertNull(
+        insertTaskLock(
+            TaskLockType.REPLACE,
+            Intervals.of("2017/2018"),
+            MEDIUM_PRIORITY,
+            tasks
+        )
+    );
+
+    // An append lock cannot be created for an overlapping interval
+    Assert.assertNull(
+        insertTaskLock(
+            TaskLockType.APPEND,
+            Intervals.of("2017-05-01/2018-05-01"),
+            MEDIUM_PRIORITY,
+            tasks
+        )
+    );
+
+    final Set<TaskLock> expectedLocks = ImmutableSet.of(
+        theLock
+    );
+    Assert.assertEquals(expectedLocks, getAllLocks(tasks));
+  }
+
+  @Test
+  public void testExclusiveLockCanRevokeAllIncompatible() throws Exception
+  {
+    final List<Task> tasks = new ArrayList<>();
+
+    // Revoked shared lock of higher priority -> revoked locks are not incompatible
+    final TaskLock sharedLock = insertTaskLock(
+        TaskLockType.SHARED,
+        Intervals.of("2016/2019"),
+        HIGH_PRIORITY,
+        tasks
+    );
+    Assert.assertNotNull(sharedLock);
+    lockbox.revokeLock(tasks.get(0).getId(), sharedLock);
+
+    // Active Exclusive lock of lower priority -> will be revoked
+    final TaskLock exclusiveLock = insertTaskLock(
+        TaskLockType.EXCLUSIVE,
+        Intervals.of("2017-01-01/2017-02-01"),
+        LOW_PRIORITY,
+        tasks
+    );
+    Assert.assertNotNull(exclusiveLock);
+    Assert.assertFalse(exclusiveLock.isRevoked());
+
+    // Active replace lock of lower priority -> will be revoked
+    final TaskLock replaceLock = insertTaskLock(
+        TaskLockType.REPLACE,
+        Intervals.of("2017-07-01/2018-01-01"),
+        LOW_PRIORITY,
+        tasks
+    );
+    Assert.assertNotNull(replaceLock);
+    Assert.assertFalse(replaceLock.isRevoked());
```

Comment:
This also seems like it should be fine, the shared lock was revoked?

Same mis-reading here.  If you are taking the time to have comments, please always make sure that they are in sync.  An incorrect comment is *significantly* worse than no comment at all as it creates confusion and encourages people to assume they know what the code is doing without actually reading it.

Reply:
Apologies, I should have elaborated further in the comment.
The tests to check compatiblity create a lock of a specific type initially and then try to see if the lock was granted or not later. 
The tests to check the revoking create a few locks that may be active initially, and are then revoked at the very end after the lock of the chosen type is granted.

Reply:
Just an over-arching comment about comments.  Comments in code are most often a negative addition to the code.  This is because they are most commonly

1) Written from the perspective of the developer who is writing the code and lacking enough context for someone who didn't write to code to know what they mean
2) Go stale after the first code adjustment because they get overlooked in code changes.

Comments should only be added when the code itself cannot be read.  When there's some extra detail in there that it would be difficult for someone to know by reading the code (there's a great example of a place where a comment is helpful in your most recent committed code where you are explaining that an append lock with a higher priority than a replace lock still doesn't revoke the replace lock).

So, when you find yourself writing comments in code please always make sure of two things

1) Is the comment really needed?  Is the code really not understandable and if it's not understandable, can it be re-arranged to be understandable
2) Assuming you've determined that a comment is actually necessary, make sure that you write it from the perspective of trying to explain something to someone who is browsing the code for the very first time, because that's the only way you will include enough context for it to be meaningful to the next person.